### PR TITLE
TreeDB: specialize sorted cached batch writes

### LINKED ISSUE #2220: Specialize TreeDB sorted and sequential cached write paths
Parent tracker: https://github.com/snissn/gomap/issues/2216

## Goal

Exploit ordered/unique input in sequential and sorted dataset write paths to avoid unnecessary hashing, sorting, latest-index rebuilds, and generic merge work.

## Why This Exists

Durable sorted/sequential write profiles show CRC/xxhash, `cmpbody`, `sort.partition_func`, `appendEntryLocked`, memtable latest-index work, and generic `setDirect` overhead. The benchmark input shape is `be8`; sorted/sequential writes should be able to take a narrower path than arbitrary random updates.

## Current Evidence

Durable 1M-key run on `mikers@192.168.0.185`:

```sh
./bin/unified-bench -dbs=treedb -keys=1000000 -profile=durable -profile-dir=$HOME/tmp/treedb_1m_durable_native_20260602_210725 -path-label=native-fastpath -format=markdown -progress=false
```

- commit: `8e89691bdaaf5bd9278a759cfb31ca9bb749d4d7`
- profile dir: `/home/mikers/tmp/treedb_1m_durable_native_20260602_210725`
- reports: `/home/mikers/tmp/treedb_1m_durable_native_20260602_210725/benchprof_results.md`, `/home/mikers/tmp/treedb_1m_durable_native_20260602_210725/insights.md`
- resolved mode: `durability=durable`, `read_integrity=verify`, cached write path with redo log on; `index_outer_leaves_in_vlog=true`; prefix/columnar/packed index optimizations disabled.
- mmap health: hit ratio `1.000000`, fallback `ReadAt=0`; this work should not focus on mmap fallback.

| Test | TreeDB ops/sec |
| --- | ---: |
| Sequential Write | 1,593,675 |
| Random Write | 1,309,137 |
| Dataset Write (Random) | 982,526 |
| Dataset Write (Sorted) | 1,184,285 |
| Batch Write | 5,834,771 |
| Batch Write (Steady) | 488,637 |
| Batch Random | 2,167,092 |
| Batch Delete | 3,359,927 |
| Batch Small Seq | 1,708,769 |
| Random Delete | 1,631,361 |
| Random Read | 1,071,324 |
| Random Read (Parallel) | 8,751,902 |
| Random Read (Parallel, Snapshot Per Key) | 1,311,219 |
| Random Read (Batch) | 3,730,882 |
| Full Scan | 4,100,139 |
| Prefix Scan | 4,035,142 |

## Branch And PR Policy

- [ ] Work must happen on a topic branch and PR; direct pushes to `main` are out of scope.
- [ ] Every PR links this issue and the durable-profile parent tracker.
- [ ] PRs must include start-phase test/benchmark plan, close-phase test evidence, close-phase benchmark evidence, and a performance regression assessment.
- [ ] Performance-sensitive changes require before/after evidence against the pre-change baseline using identical commands, fixture size, hardware, and environment.
- [ ] Material regressions in throughput, allocation, storage footprint, latency, syscall/block/mutex counters, or relevant TreeDB stats are blocking until optimized away or explicitly accepted by the coordinator/user.
- [ ] Request Codex/Copilot/CodeRabbit reviews only after the PR is mature: coherent code, tests/benchmark evidence or rationale, no known blockers, and latest-head CI running or green.
- [ ] Latest-head CI is required for mergeability; stale green checks do not count.

## Standard Benchmark Matrix

Use focused commands for local iteration and at least one relevant 1M-key durable before/after comparison on `mikers@192.168.0.185` when claiming a performance improvement:

```sh
OUT=$HOME/tmp/<issue-specific-run>
./bin/unified-bench -dbs=treedb -keys=1000000 -profile=durable -profile-dir="$OUT" -path-label=native-fastpath -format=markdown -progress=false -test=<focused-tests>
./bin/benchprof -profiles-dir "$OUT"
```

Record ops/sec, allocation evidence (`B/op`, `allocs/op` or pprof alloc objects/space), relevant pprof top entries, and TreeDB counters. If a microbenchmark is added, report `ns/op`, `ops/sec`, `B/op`, and `allocs/op`.

## Dependencies

- Depends on the non-alloc key-lock issue for cleaner allocation attribution.
- Coordinate with WAL/checksum work to avoid overlapping write-path refactors.

## Scope

- Detect monotonic/sorted unique batches or direct writes where safe.
- Avoid redundant sort/latest-index rebuild work for ordered unique key streams.
- Consider key64-specialized fast paths only when they preserve generic correctness and fall back safely.

## Non-Goals

- Do not special-case benchmark-only code paths in a way application writes cannot use.
- Do not bypass WAL/checkpoint/snapshot correctness.

### M0. Invariant proof
- [ ] Document the exact ordered-input invariant and fallback behavior.
- [ ] Add tests for sorted unique, sorted duplicate, and out-of-order fallback cases.

### M1. Implementation
- [ ] Implement a minimal ordered fast path in the cached write/memtable layer.
- [ ] Keep generic random-write behavior unchanged.

### M2. Evidence
- [ ] Run focused durable `sequential_write,dataset_write_sorted,batch_small_seq` before/after.
- [ ] Verify random write/delete do not regress materially.

## Required Tests

- [ ] Focused TreeDB/caching/memtable sorted-write tests.
- [ ] `GOWORK=off go test ./TreeDB/... -count=1` if memtable invariants change.

## Completion Criteria

- [ ] Ordered workloads show reduced CPU/allocation profile cost.
- [ ] Fallback and duplicate-key semantics are covered by tests.

### PR DESCRIPTION
**Pull request by snissn**
## Summary

Fixes #2220. Part of #2216.

This adds a narrow cached-write fast path for already-sorted unique batch input:

- adds internal `memtable.CopySortedBatchApplier` for sorted batches that cannot transfer key ownership;
- lets cached batch writes use the sorted applier for append-only/hash-sorted memtables when `Batch` has proven strictly increasing keys;
- keeps duplicate, out-of-order, mixed unsupported, and existing-key-overlap cases on the generic path;
- preserves generic/direct `AppendOnly.appendEntryLocked` behavior after benchmarking found a direct-write regression in an earlier draft;
- adds tests for sorted unique, duplicate fallback, out-of-order fallback, SetView non-borrowing, pointer inline policy, and key64 monotonic ordering.

No WAL/value-log append/sync ordering changes. No on-disk format changes.

## Ordered-input invariant

Fast path is used only when cached `Batch` has tracked a strictly increasing key stream (`bytes.Compare(prev, key) < 0`). Per-shard subsequences preserve that order. Memtable appliers also require the run to append after the current max key; otherwise they fall back to generic latest-wins append/update handling.

Value ownership:

- keys are always copied into memtable-owned storage for `CopySortedBatchApplier`;
- values may be borrowed only when the cached batch arena can be retained, and the applier returns whether it borrowed;
- `SetView` disables value borrowing and is covered by test.

## Tests

```sh
GOWORK=off go test ./TreeDB/internal/memtable ./TreeDB/caching -run 'TestAppendOnlyApplyCopySortedBatchTrusted|TestAppendOnlyKey64MonotonicOrderingMatchesLexicographic|TestHashSortedApply|TestBatchWriteSorted|TestBatchWriteOutOfOrder|TestBatchSetOps_StreamEligible' -count=1
GOWORK=off go test ./TreeDB/internal/memtable ./TreeDB/caching -count=1
GOWORK=off go test ./TreeDB/... -count=1
```

## Benchmark evidence

Host: `mikers@192.168.0.185`, 1M keys, durable profile, native fast path.

Command for each before/after run:

```sh
./bin/unified-bench -dbs=treedb -keys=1000000 -profile=durable \
  -profile-dir="$OUT" -path-label=native-fastpath -format=markdown -progress=false \
  -test=sequential_write,dataset_write_sorted,batch_small_seq,random_write,random_delete
```

Artifacts root: `/home/mikers/tmp/gomap_2220_bench_20260602_231318`

- baseline dirs: `profiles_before_1m`, `profiles_before_1m_r2`, `profiles_before_1m_after_latest`
- after dirs: `profiles_after_final_1m`, `profiles_after_final_1m_r2`, `profiles_after_latest_1m`

Three-run average ops/sec:

| Test | Before avg | After avg | Delta |
| --- | ---: | ---: | ---: |
| Sequential Write | 2,154,103 | 2,132,675 | -0.99% |
| Dataset Write (Sorted) | 1,699,818 | 1,693,900 | -0.35% |
| Batch Small Seq | 2,370,185 | 2,769,789 | +16.86% |
| Random Write | 1,645,326 | 1,675,626 | +1.84% |
| Random Delete | 2,021,956 | 1,986,621 | -1.75% |

Latest-head paired sanity (`profiles_after_latest_1m` then `profiles_before_1m_after_latest`):

| Test | Before | After @ `3e17b7c2` | Delta |
| --- | ---: | ---: | ---: |
| Sequential Write | 2,094,476 | 2,084,264 | -0.49% |
| Dataset Write (Sorted) | 1,629,792 | 1,638,563 | +0.54% |
| Batch Small Seq | 2,361,206 | 2,743,878 | +16.20% |
| Random Write | 1,669,100 | 1,686,287 | +1.03% |
| Random Delete | 2,062,142 | 1,978,414 | -4.06% |

Assessment: targeted small sorted batches improve materially. Direct Set-style sorted/sequential numbers are noisy and are not claimed as improved; final code keeps the generic direct append path unchanged relative to `origin/main`. Random write/delete sanity is within observed run-to-run noise on the multi-run average; no WAL, mmap fallback, storage footprint, or value-log durability counters regressed materially.

## Review / CI / AI status

- Internal high-thinking review: PASS, no blocking findings.
- CI: pending for PR head.
- AI reviews: not requested yet; will request after PR is mature and CI is running/green per tracker policy.

## Release Notes

* **New Features**
  * Introduced optimized fast path for sorted, unique batch writes.
  * Added support for copy-based batch operations when external data cannot transfer ownership.

* **Improvements**
  * Enhanced streaming batch write path to maintain sorted-run accounting across all execution paths.

* **Tests**
  * Added comprehensive test coverage for sorted batch write optimization paths and fallback scenarios.

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/internal/memtable/hash_sorted.go [hunk 1]
```diff
@@ -217,6 +217,45 @@ func (m *HashSorted) ApplyStealSortedBatchTrusted(entries []batchpkg.Entry, onKe
...
+			if keyStored, n, ok := m.setEntryNewCopyNoChunkLocked(op, storeInlinePtrValues); ok {
+				keys = append(keys, keyStored)
+				keyBytes += n
+			}
+			if onKey != nil {
+				onKey(op.Key)
+			}
+		}
+		if chunk, seq := m.noteNewKeysBatchLocked(keys, keyBytes); seq != 0 {
+			chunks = append(chunks, hashSortedIndexWork{mt: m, seq: seq, keys: chunk, sorted: true})
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/memtable/hash_sorted.go]: Fixed in b77eaec2d: HashSorted now tracks whether pending key chunks remain sorted. Copy/steal sorted-batch fast paths pass through the computed chunk sortedness, so a sorted=true enqueue only happens for a pure/known-sorted pending chunk; mixed prior unsorted pending keys are enqueued with sorted=false. Added targeted mixed-pending and consecutive-sorted pending tests.

File: TreeDB/caching/sorted_write_fastpath_test.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	if err := b.Write(); err != nil {
+		t.Fatalf("Write: %v", err)
+	}
+
+	if rec.calls != 0 {
+		t.Fatalf("copy sorted fast path used for out-of-order batch: calls=%d", rec.calls)
+	}
+	requireCachedValue(t, db, []byte("a"), []byte("va"))
+	requireCachedValue(t, db, []byte("b"), []byte("vb"))
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/sorted_write_fastpath_test.go:258]: Fixed in b77eaec2d: added TestBatchWriteSortedOverlappingExistingKeyFallsBackToGeneric covering sorted unique batch overlap with an existing memtable key and verifying latest value semantics plus no recorded ordered-append fast-path call.

File: TreeDB/internal/memtable/append_only.go [hunk 3]
```diff
@@ -745,6 +745,60 @@ func (m *AppendOnly) appendEntryLocked(key, value []byte, ptr page.ValuePtr, fla
...
+		ent.value = nil
+		ent.valueOwned = false
+		ent.ptr = page.ValuePtr{}
+	}
+	k := appendOnlyEntryKey(ent)
+	m.sizeBytes += int64(len(k) + entryValueSize(flags, ent.value))
+	m.lastIdx = idx
+	m.hasLast = true
+	return ent
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/memtable/append_only.go:758]: Fixed in b77eaec2d: extracted appendEntryCoreLocked for AppendOnly allocation/key-value copy/size accounting, leaving appendEntryLocked and appendEntryTrustedOrderedLocked to handle only their differing order/index post-processing.

File: TreeDB/internal/memtable/hash_sorted.go [hunk 4]
```diff
@@ -933,6 +972,44 @@ func (m *HashSorted) setEntryLocked(key, value []byte, ptr page.ValuePtr, flags
...
+		return nil, page.ValuePtr{}, node.FlagTombstone
+	case op.IsPtr:
+		if storeInlinePtrValues {
+			return op.Value, op.ValuePtr, node.FlagPointer
+		}
+		return nil, op.ValuePtr, node.FlagPointer
+	default:
+		return op.Value, page.ValuePtr{}, node.FlagInline
+	}
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/memtable/hash_sorted.go:1003]: Fixed in b77eaec2d: added shared batchEntryPayload helper and made AppendOnly/HashSorted payload wrappers delegate to it.

**Comment by snissn** [comment]
Update after coordinator/Codex review:

- Fixed the HashSorted pending-key chunk ordering bug in `b77eaec2d`.
  - `HashSorted` now tracks whether pending keys are known sorted.
  - Sorted-batch append paths enqueue `sorted=true` only for pure/known-sorted chunks.
  - Mixed chunks containing prior generic/non-append pending keys enqueue `sorted=false`, so the indexer sorts them.
- Added targeted coverage:
  - mixed generic pending keys + sorted copy batch crossing the seal threshold is **not** marked sorted;
  - consecutive sorted pending batches still produce a sorted chunk;
  - sorted unique batch overlapping an existing cached key falls back and preserves latest value semantics.
- Addressed CodeRabbit quick wins:
  - overlap fallback test added;
  - AppendOnly append allocation/copy/size accounting deduplicated via `appendEntryCoreLocked`;
  - shared `batchEntryPayload` helper for AppendOnly/HashSorted.

Local validation on latest head `b77eaec2d`:

```sh
GOWORK=off go test ./TreeDB/internal/memtable ./TreeDB/caching -run 'TestHashSortedApplyCopySortedBatchTrusted|TestAppendOnlyApplyCopySortedBatchTrusted|TestAppendOnlyKey64MonotonicOrderingMatchesLexicographic|TestHashSortedApply|TestBatchWriteSorted|TestBatchWriteOutOfOrder|TestBatchSetOps_StreamEligible' -count=1
GOWORK=off go test ./TreeDB/internal/memtable ./TreeDB/caching -count=1
GOWORK=off go test ./TreeDB/... -count=1
```

All passed locally. Latest-head CI is running; will re-request AI reviews after CI settles.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review
